### PR TITLE
Improve numerical stability of 3D projection

### PR DESCRIPTION
**Pull request by hborchardt**
When the position vector has large values that are cancelled out by large values in the model matrix, it is more numerically stable to first multiply the position by the model matrix instead of multiplying the matrices together first.

This should always yield equivalent or more accurate results.

Use forks of the gl-vis libraries that use the problematic pattern.

Helps with https://github.com/plotly/plotly.js/issues/3306
See https://github.com/gl-vis/gl-axes3d/pull/23

### CODE REVIEW

**Comment by archmoj** [comment]
Thanks for the PR.
Please download `baselines.tar` from https://app.circleci.com/pipelines/github/plotly/plotly.js/10465/workflows/5b8aa5a0-5c2e-4237-b231-604b94893e2b/jobs/229556/artifacts
Then extract the files and replace them in `test/image/baselines`; commit and push.

**Comment by archmoj** [comment]
Please add `gl3d_mesh3d_surface3d_scatter3d_line3d_error3d_log_reversed_ranges` to this list:
https://github.com/plotly/plotly.js/blob/8b09892dcfbb859067794279ce9ffa9e2df95dd4/test/image/compare_pixels_test.js#L176-L181 which would help `test-baselines-virtual-webgl` pass.

File: stackgl_modules/package.json [hunk 1]
```diff
@@ -58,5 +58,16 @@
     "node-polyfill-webpack-plugin": "3.0.0",
     "webpack": "5.90.3",
     "webpack-cli": "5.1.4"
+  },
+  "overrides": {
```
Comment (archmoj) [inline stackgl_modules/package.json:62]: I don't think we need to have `overrides`.
Could you please revert the changes to `package.json` and `packages-lock.json` then 
inside `stackgl_modules` install these:
```sh
npm install https://github.com/hborchardt/gl-axes3d#83caafc058b2d60b6006e26432967c411c125fe1
npm install https://github.com/hborchardt/gl-cone3d#e9728cd068e0dfe11b92c5785974cd463f78c01c
npm install https://github.com/hborchardt/gl-error3d#ee55620d17f3cc7dfb104823c904f5070aba79ed
npm install https://github.com/hborchardt/gl-line3d#27123949f65e2d8deab6a6e131030073ed9cdf02
npm install https://github.com/hborchardt/gl-mesh3d#be2c7fcb68e70e001bbfb8021380d0d7ac75b1c9
npm install https://github.com/hborchardt/gl-scatter3d#086177000f405bc72f171350f9c1efa6eb523e29
npm install https://github.com/hborchardt/gl-streamtube3d#29c914cc0d88fc5dffce62dcd8169f38115d078b
npm install https://github.com/hborchardt/gl-surface3d#0728e33c0296c602b147e8f1ea8f1bfe7040c61e
```

Comment (hborchardt) [reply stackgl_modules/package.json:62]: Okay, but in that case I need to add 
`npm install https://github.com/hborchardt/gl-spikes3d#9ad06b2dd48fa19d2faa54e5645c2e4bf5e6157c`
as well, because that contains the pattern but is only indirectly depended on via e.g. gl-mesh3d

File: stackgl_modules/package.json [hunk 2]
```diff
@@ -13,21 +13,22 @@
     "box-intersect": "plotly/box-intersect#v1.1.0",
     "convex-hull": "^1.0.3",
     "delaunay-triangulate": "^1.1.6",
-    "gl-axes3d": "^1.7.0",
-    "gl-cone3d": "^1.6.0",
-    "gl-error3d": "^1.0.16",
+    "gl-axes3d": "github:hborchardt/gl-axes3d#83caafc058b2d60b6006e26432967c411c125fe1",
```
Comment (archmoj) [inline stackgl_modules/package.json:16]: Please paste a link to the `gl-axes3d` pull request here.

Comment (hborchardt) [reply stackgl_modules/package.json:16]: https://github.com/gl-vis/gl-axes3d/pull/23

File: stackgl_modules/package.json [hunk 3]
```diff
@@ -13,21 +13,22 @@
...
+    "gl-cone3d": "github:hborchardt/gl-cone3d#e9728cd068e0dfe11b92c5785974cd463f78c01c",
```
Comment (archmoj) [inline stackgl_modules/package.json:17]: Please open pull requests on all `gl-vis/*` dependencies  and paste the links for every line change in `package.json` file. Thank you!

Comment (hborchardt) [reply stackgl_modules/package.json:17]: See https://github.com/gl-vis/gl-cone3d/pull/21

File: stackgl_modules/package.json [hunk 4]
```diff
@@ -13,21 +13,22 @@
...
+    "gl-error3d": "github:hborchardt/gl-error3d#ee55620d17f3cc7dfb104823c904f5070aba79ed",
```
Comment (hborchardt) [inline stackgl_modules/package.json:18]: See https://github.com/gl-vis/gl-error3d/pull/12

File: stackgl_modules/package.json [hunk 5]
```diff
@@ -13,21 +13,22 @@
...
     "gl-heatmap2d": "^1.1.1",
-    "gl-line3d": "1.2.1",
-    "gl-mesh3d": "^2.3.1",
+    "gl-line3d": "github:hborchardt/gl-line3d#27123949f65e2d8deab6a6e131030073ed9cdf02",
```
Comment (hborchardt) [inline stackgl_modules/package.json:20]: https://github.com/gl-vis/gl-line3d/pull/15

File: stackgl_modules/package.json [hunk 6]
```diff
@@ -13,21 +13,22 @@
...
+    "gl-mesh3d": "github:hborchardt/gl-mesh3d#be2c7fcb68e70e001bbfb8021380d0d7ac75b1c9",
```
Comment (hborchardt) [inline stackgl_modules/package.json:21]: See https://github.com/gl-vis/gl-mesh3d/pull/23

File: stackgl_modules/package.json [hunk 7]
```diff
@@ -13,21 +13,22 @@
...
     "gl-plot2d": "^1.5.0",
     "gl-plot3d": "^2.4.7",
     "gl-pointcloud2d": "^1.0.3",
-    "gl-scatter3d": "^1.4.0",
+    "gl-scatter3d": "github:hborchardt/gl-scatter3d#086177000f405bc72f171350f9c1efa6eb523e29",
```
Comment (hborchardt) [inline stackgl_modules/package.json:25]: See https://github.com/gl-vis/gl-scatter3d/pull/23

File: stackgl_modules/package.json [hunk 8]
```diff
@@ -13,21 +13,22 @@
...
     "gl-select-box": "^1.0.4",
     "gl-shader": "4.3.1",
     "gl-spikes2d": "^1.0.2",
-    "gl-streamtube3d": "^1.4.1",
-    "gl-surface3d": "^1.6.0",
+    "gl-spikes3d": "github:hborchardt/gl-spikes3d#9ad06b2dd48fa19d2faa54e5645c2e4bf5e6157c",
```
Comment (hborchardt) [inline stackgl_modules/package.json:29]: See https://github.com/gl-vis/gl-spikes3d/pull/6

File: stackgl_modules/package.json [hunk 9]
```diff
@@ -13,21 +13,22 @@
...
+    "gl-streamtube3d": "github:hborchardt/gl-streamtube3d#29c914cc0d88fc5dffce62dcd8169f38115d078b",
```
Comment (hborchardt) [inline stackgl_modules/package.json:30]: See https://github.com/gl-vis/gl-streamtube3d/pull/13

File: stackgl_modules/package.json [hunk 10]
```diff
@@ -13,21 +13,22 @@
...
+    "gl-surface3d": "github:hborchardt/gl-surface3d#0728e33c0296c602b147e8f1ea8f1bfe7040c61e",
```
Comment (hborchardt) [inline stackgl_modules/package.json:31]: See https://github.com/gl-vis/gl-surface3d/pull/20

**Comment by archmoj** [comment]
Looking at the diffs of baselines (https://app.circleci.com/pipelines/github/plotly/plotly.js/10465/workflows/5b8aa5a0-5c2e-4237-b231-604b94893e2b/jobs/229556/artifacts), the changes are not remarkable on most of except the one ussing dates and larger numbers which is a good sign of fixing the bug.
For reference I pasted the diff of `gl3d_world-cals` mock below.
diff-gl3d_world-cals

Would you please paste two short videos interacting with `gl3d_world-cals` or other mock showing before and after effects?

**Comment by archmoj** [comment]
For the before vs after video you may fork and tweak this [codepen](https://codepen.io/MojtabaSamimi/pen/MWdaggx?editors=0010).

File: stackgl_modules/package.json [hunk 11]
```diff
@@ -13,21 +13,22 @@
     "box-intersect": "plotly/box-intersect#v1.1.0",
     "convex-hull": "^1.0.3",
     "delaunay-triangulate": "^1.1.6",
-    "gl-axes3d": "^1.7.0",
-    "gl-cone3d": "^1.6.0",
-    "gl-error3d": "^1.0.16",
+    "gl-axes3d": "github:hborchardt/gl-axes3d#fcc1c5c49db3de6445020a63a5d76403ec5a2b98",
```
Comment (hborchardt) [inline stackgl_modules/package.json:16]: See https://github.com/gl-vis/gl-axes3d/pull/23

**Comment by hborchardt** [comment]
For `gl3d_world-cals`, I scaled the z axis values to be within a few hours, to make the problem apparent.
On `master`:
[old.webm](https://github.com/plotly/plotly.js/assets/66408901/32433a3b-626e-476a-a6b2-e7ccc8e9d8af)
On this branch:
[new.webm](https://github.com/plotly/plotly.js/assets/66408901/6ab1124c-0e4f-49e2-8062-6419d507a8ea)

Unfortunately this did not fix the jagged edges of the surface plot - would have been nice.

File: stackgl_modules/package.json (hunk 11 above)
Comment (archmoj) [reply stackgl_modules/package.json:16]: Please install the patch.
```
npm i gl-axes3d@1.7.1
```

File: stackgl_modules/package.json (hunk 10 above)
Comment (archmoj) [reply stackgl_modules/package.json:31]: Please install the patch.
```
npm i gl-surface3d@1.6.1
```

File: stackgl_modules/package.json (hunk 9 above)
Comment (archmoj) [reply stackgl_modules/package.json:30]: Now: `gl-streamtube3d@1.4.2`

File: stackgl_modules/package.json (hunk 7 above)
Comment (archmoj) [reply stackgl_modules/package.json:25]: Now: `gl-scatter3d@1.4.1`

File: stackgl_modules/package.json (hunk 6 above)
Comment (archmoj) [reply stackgl_modules/package.json:21]: Now: `gl-mesh3d@2.3.2`

File: stackgl_modules/package.json (hunk 5 above)
Comment (archmoj) [reply stackgl_modules/package.json:20]: Now: `gl-line3d@1.2.2`

File: stackgl_modules/package.json (hunk 4 above)
Comment (archmoj) [reply stackgl_modules/package.json:18]: Now: `gl-error3d@1.0.17`

File: stackgl_modules/package.json (hunk 3 above)
Comment (archmoj) [reply stackgl_modules/package.json:17]: Now: `gl-cone3d@1.6.1`

**Comment by archmoj** [comment]
@hborchardt , please install new patches of `gl-vis` modules and we should be good to go :rocket:

File: draftlogs/6998_fix.md [hunk 12]
```diff
@@ -0,0 +1 @@
+- Fix numerical instability in 3D plots [[6998](https://github.com/plotly/plotly.js/pull/6998)]
```
Comment (archmoj) [inline draftlogs/6998_fix.md:1]:
```suggestion
 - Fix numerical instability in 3D plots [[6998](https://github.com/plotly/plotly.js/pull/6998)],
   with thanks to @hborchardt for the contribution!
```

**Comment by archmoj** [comment]
Plotly.js v2.33.0 is out: https://github.com/plotly/plotly.js/releases/tag/v2.33.0
Thanks for your contribution.
